interfaces/builtin/netlink_protocol.go

```diff
@@ -0,0 +1,92 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package builtin
+
+import (
+	"fmt"
+
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/snap"
+)
+
+const netlinkProtocolSummary = `allows communication through the kernel custom netlink protocol`
+
+const netlinkProtocolBaseDeclarationSlots = `
+  netlink-protocol:
+    allow-installation:
+      slot-snap-type:
+        - core
+        - gadget
```

Comment:
why is the gadget meant to provide this slot? Is the netlink protocol number tied to physical hardware? If it's not then I would prefer this just be a normal slot that is provided by an app

Reply:
This PR is designed to support arbitrary netlink families / protocol numbers - the kernel has a number of pre-defined ones that are supported (https://man7.org/linux/man-pages/man7/netlink.7.html) but some users/customers will implement their own custom protocols as separate drivers and so require a way to use these from userspace. As these protocol numbers then are driver specific it makes sense for these slots to be defined by the gadget snap IMO.

Reply:
I understand there is a customer specific component to netlink, but having customer specific code does not explain why it should come from a gadget. Typically we try to make gadget provided things tied to hardware bits like gpio, i2c, etc.

We have other such interfaces which are just app provided slots and not system provided slots.

Reply:
as far as I understand the protocol here will be custom and specific to a device (family), so I think the gadget is a better place for that, it also a declaration that the device has a device with such a protocol.

Reply:
There is no perfect place for this, kernel snap would be the correct one. But it definitely more gadget than app snap, as this is implemented by the kernel driver.
As @pedronis said, it's something which is specific to a family of devices.

Reply:
providing by the kernel makes the most sense to me here, not sure if we support that, but I suppose in lieu of being able to do kernel provided slots a gadget provided slot is closer